### PR TITLE
Map unmatched `node_modules` files

### PR DESCRIPTION
**Pull request by keshav-space**
Map the 3rd-party files inside `node_modules` that couldn't be matched to PurlDB due to slight modifications.

### CODE REVIEW

**Comment by tdruez** [comment]
@keshav-space is this PR complete and ready to be reviewed?
If not, please provide the latest status and a list of remaining tasks.

File: scanpipe/pipes/js.py [hunk 1]
```diff
@@ -204,3 +206,21 @@ def get_js_map_basename_and_extension(filename):
         if filename.endswith(ext):
             basename = filename[: -len(ext)]
             return basename, ext
+
+
+def get_purl_from_node_module(node_module_directory):
+    """Return PURL for given a `node_modules` package directory."""
```
Comment (pombredanne) [inline scanpipe/pipes/js.py:212]: Do you have a test for this?

File: scanpipe/pipes/purldb.py [hunk 2]
```diff
@@ -117,3 +117,13 @@ def match_resource(sha1_list, timeout=None, api_url=PURLDB_API_URL):
     if response and response.get("count"):
         packages = response["results"]
         return packages
+
+
+def fetch_package(purl, timeout=None, api_url=PURLDB_API_URL):
+    """Fetch package data for the PURL."""
+    payload = {"purl": purl}
+    response = request_get(url=f"{api_url}packages/", payload=payload, timeout=timeout)
```
Comment (pombredanne) [inline scanpipe/pipes/purldb.py:125]: We did not have one before?

**Review by pombredanne — CHANGES_REQUESTED** [review]
See some nits for your review.

**Comment by pombredanne** [comment]
Also can you elaborate in the body of the PR what this is doing?

File: scanpipe/pipes/purldb.py (hunk 2 above)
Comment (keshav-space) [reply scanpipe/pipes/purldb.py:135]: We didn't have one for getting package using PURL

File: scanpipe/pipes/js.py (hunk 1 above)
Comment (keshav-space) [reply scanpipe/pipes/js.py:212]: Added test for this.

File: scanpipe/pipes/d2d.py [hunk 3]
```diff
@@ -840,3 +840,102 @@ def _map_javascript_path_resource(
...
+        type=purl.type,
+        namespace="" if not purl.namespace else purl.namespace,
+        name=purl.name,
+        version=purl.version,
+    ).first()
+
+    if package:
+        package.add_resources(matched)
+    else:
+        if results := purldb.fetch_package(purl=str(purl)):
```
Comment (tdruez) [inline scanpipe/pipes/d2d.py:927]: I'm not sure about depending on a purldb call deep within a d2d pipe.
@pombredanne what's your take on this architecture? We've only implemented pipes and steps dedicated to purldb so far.

Comment (pombredanne) [reply scanpipe/pipes/d2d.py:1076]: Or would you want to have a separate step?
@tdruez what alternative approach would you see?

File: scanpipe/pipes/js.py [hunk 4]
```diff
@@ -261,3 +263,21 @@ def map_related_files(to_resources, to_resource, from_resource, map_type, extra_
         match.update(status=flag.MAPPED)
 
     return len(transpiled)
+
+
+def get_purl_from_node_module(node_module_directory):
+    """Return PURL for given a `node_modules` package directory."""
```
Comment (pombredanne) [inline scanpipe/pipes/js.py:269]: Can we have some examples of the paths in the docstring?

**Review by pombredanne — CHANGES_REQUESTED** [review]
Here is some extre feedback.

**Comment by keshav-space** [comment]
closing this in favour of https://github.com/nexB/scancode.io/pull/881

**Pull request closed without merging**